#58: VeLa variables may undesirably change type

dbenn opened the pull request:
Fixed by converting new value's type to the type of the existing binding or throwing an exception; added comments and deprecated a function.

yorkpf commented:
David

I have started to look at this.

I guess that “reviewing” involves looking at #58 (to understand the nature of the issue) and also looking at #118 (to understand the proposed solution).

The 3 links below, marked with an “M” appear to be the file changes (+ and -) proposed, but point to different portions of the same document that comes up when you click on “Files Changed” on the github site. I understand I need to look at these but what about the two links below them (one ends in .patch and one in .diff) - what do I do with these?

Paul

> On 14 Feb 2021, at 5:16 pm, dbenn <notifications@github.com <mailto:notifications@github.com>> wrote:
> 
> Fixed by converting new value's type to the type of the existing binding or throwing an exception; added comments and deprecated a function.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/AAVSO/VStar/pull/118 <https://github.com/AAVSO/VStar/pull/118>
> Commit Summary
> 
> #58: fixed by converting new value's type to the type of the existing binding or throwing an exception; added comments and deprecated a function
> File Changes
> 
> M src/org/aavso/tools/vstar/vela/VeLaEnvironment.java <https://github.com/AAVSO/VStar/pull/118/files#diff-19fa1031c4a76bc7f014bdb41fa5e0a08b2cefc0555141d251c27b38316b5fa5> (1)
> M src/org/aavso/tools/vstar/vela/VeLaInterpreter.java <https://github.com/AAVSO/VStar/pull/118/files#diff-92cbcefa602b71df2af46793945d5ec78708ab7b47bbc04b25eeadf62755a3d4> (24)
> M test/org/aavso/tools/vstar/vela/VeLaTest.java <https://github.com/AAVSO/VStar/pull/118/files#diff-e96d01ac0c782e6883730d6bfdee1a2396b03b01969c2a96e412ce5f3f85e045> (20)
> Patch Links:
> 
> https://github.com/AAVSO/VStar/pull/118.patch <https://github.com/AAVSO/VStar/pull/118.patch>
> https://github.com/AAVSO/VStar/pull/118.diff <https://github.com/AAVSO/VStar/pull/118.diff>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/AAVSO/VStar/pull/118>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASK654CPGA435KDHS77RYALS652DRANCNFSM4XS6NTVQ>.
>

dbenn commented:
That’s all correct. You don’t need to worry about the diff and patch files. They can be used with particular tools to modify source code.

David

> On 15 Feb 2021, at 16:06, Paul <notifications@github.com> wrote:
> 
> 
> David
> 
> I have started to look at this.
> 
> I guess that “reviewing” involves looking at #58 (to understand the nature of the issue) and also looking at #118 (to understand the proposed solution).
> 
> The 3 links below, marked with an “M” appear to be the file changes (+ and -) proposed, but point to different portions of the same document that comes up when you click on “Files Changed” on the github site. I understand I need to look at these but what about the two links below them (one ends in .patch and one in .diff) - what do I do with these?
> 
> Paul
> 
> > On 14 Feb 2021, at 5:16 pm, dbenn <notifications@github.com <mailto:notifications@github.com>> wrote:
> > 
> > Fixed by converting new value's type to the type of the existing binding or throwing an exception; added comments and deprecated a function.
> > 
> > You can view, comment on, or merge this pull request online at:
> > 
> > https://github.com/AAVSO/VStar/pull/118 <https://github.com/AAVSO/VStar/pull/118>
> > Commit Summary
> > 
> > #58: fixed by converting new value's type to the type of the existing binding or throwing an exception; added comments and deprecated a function
> > File Changes
> > 
> > M src/org/aavso/tools/vstar/vela/VeLaEnvironment.java <https://github.com/AAVSO/VStar/pull/118/files#diff-19fa1031c4a76bc7f014bdb41fa5e0a08b2cefc0555141d251c27b38316b5fa5> (1)
> > M src/org/aavso/tools/vstar/vela/VeLaInterpreter.java <https://github.com/AAVSO/VStar/pull/118/files#diff-92cbcefa602b71df2af46793945d5ec78708ab7b47bbc04b25eeadf62755a3d4> (24)
> > M test/org/aavso/tools/vstar/vela/VeLaTest.java <https://github.com/AAVSO/VStar/pull/118/files#diff-e96d01ac0c782e6883730d6bfdee1a2396b03b01969c2a96e412ce5f3f85e045> (20)
> > Patch Links:
> > 
> > https://github.com/AAVSO/VStar/pull/118.patch <https://github.com/AAVSO/VStar/pull/118.patch>
> > https://github.com/AAVSO/VStar/pull/118.diff <https://github.com/AAVSO/VStar/pull/118.diff>
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub <https://github.com/AAVSO/VStar/pull/118>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASK654CPGA435KDHS77RYALS652DRANCNFSM4XS6NTVQ>.
> > 
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub <https://github.com/AAVSO/VStar/pull/118#issuecomment-778955116>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA7MZLDCBELPS4GMBL63VIDS7CXHTANCNFSM4XS6NTVQ>.
>

dbenn commented:
@mpyat2 any comments on this one?

mpyat2 commented:
@dbenn no comment at this time.

yorkpf commented:
David

A trivial question: How do you pronounce" Vela” - to rhyme with “wheeler” or “weller”?

More importantly, what was the motivation for developing VeLa? I have read the material at github.com/AAVSO/VStar/wiki/VeLa <http://github.com/AAVSO/VStar/wiki/VeLa>, but I am far from clear that a Java construct of some kind could not achieve the same thing.

Regards
Paul

> On 20 Feb 2021, at 3:55 pm, dbenn <notifications@github.com> wrote:
> 
> 
> Merged #118 <https://github.com/AAVSO/VStar/pull/118> into master.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/AAVSO/VStar/pull/118#event-4354928977>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASK654AQNLJFKG7OEN3WDDTS75FF5ANCNFSM4XS6NTVQ>.
>

dbenn commented:
Paul

To rhyme with wheeler. As I thought about the acronym, it started to look astronomical after the Vela pulsar, SN remnant, and constellation.

The motivation was to create first an expression language that could be used where numbers were used in dialogs (e.g. create a phase plot or model with a period multiplied by 2 or halved, subtract ~10 years from a JD in AID load, e.g. 2458535.5-100*365.25), then to create filters, and then the other items in the list on that page emerged. Not on that list is: scripting VStar itself as an alternative to JavaScript.

But that is not the question you are asking, right? Why create a new language?

I needed something tightly integrated with VStar. I could have made that language Java and called the Java compiler from within VStar (easy enough), but that language is not suitable for simple expressions or scripting nor would many people wish to use it. I had already added JavaScript as a scripting language through Tool -> Run Script... but the JavaScript implementation available in Java changed a couple of times and I didn't want to be at the whim of Sun/Oracle. JavaScript is also too dynamically typed for my taste and I wanted to exclude as many errors as possible in this context. I could have chosen some other scripting engine language such as JRuby or Python but this engine is intended more for scripting something from the outside. As it turns out, I can use VeLa for that too. There were probably other options but I decided in the end to create something (a domain specific language) that would integrate with VStar in the way I wanted, allow localised numbers etc. I had the experience to do that. 

Why a new language, rather than an implementation for a suitably well typed existing language? The answer is less clear and more along the lines of: because I wanted to and I could. 

Maybe I should add some of this detail to the wiki page?

I'd like to hear more about your thought that "...a Java construct of some kind could not achieve the same thing."

Thanks for questioning my approach. It's fine.

David

yorkpf commented:
Thanks for that reply, David.

I can certainly understand “because I wanted to and I could”. Yes, I can definitely relate to that :-)

The explanation you gave is the gist of what I was looking for:  "that language [Java] is not suitable for simple expressions or scripting nor would many people wish to use it. I had already added JavaScript as a scripting language through Tool -> Run Script... but the JavaScript implementation available in Java changed a couple of times and I didn't want to be at the whim of Sun/Oracle. JavaScript is also too dynamically typed for my taste and I wanted to exclude as many errors as possible in this context. I could have chosen some other scripting engine language such as JRuby or Python but this engine is intended more for scripting something from the outside".

Re my comments (below) to the effect that plain Java ought to be able to achieve the same thing, that was merely speculative - I was just thinking out loud. My knowledge of Java is not deep enough to even hazard an opinion on that one. I was just curious as to your intentions.

BTW, I suppose you know that “vela” in Latin means “sails”, ie. the sails of a sailing ship?

Thanks.
Paul

> On 21 Feb 2021, at 9:43 am, dbenn <notifications@github.com> wrote:
> 
> I'd like to hear more about your thought that "...a Java construct of some kind could not achieve the same thing."

dbenn commented:
Thanks Paul.

I was aware of the sails/ship connection but had forgotten. VeLa puts the wind in your sails? ;) Vela, Carina, Crux, Centaurus together make up a very rich and beautiful part of the sky.

I appreciate my perspective and motives being challenged. Keeps me on my toes.

David

dbenn commented:
Added a brief excerpt of the foregoing to the wiki Intro section.

yorkpf commented:
Wasn’t so much a challenge, just an enquiry :-)

> On 21 Feb 2021, at 4:06 pm, dbenn <notifications@github.com> wrote:
> 
> 
> Thanks Paul.
> 
> I was aware of the sails/ship connection but had forgotten. VeLa puts the wind in your sails? ;) Vela, Carina, Crux, Centaurus together make up a very rich and beautiful part of the sky.
> 
> I appreciate my perspective and motives being challenged. Keeps me on my toes.
> 
> David
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/AAVSO/VStar/pull/118#issuecomment-782806318>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASK654C3UTOTAEX5ONK6KYLTACPEZANCNFSM4XS6NTVQ>.
>
